agora/Agora/Proposal/Time.hs

```diff
@@ -388,74 +389,108 @@ currentProposalTime = phoistAcyclic $
-      pisCurrentTimeWithin # s # (s + f.draftTime + f.votingTime)
+pisWithin :: forall (s :: S). Term s (PTimingRelation :--> PBool)
+pisWithin = phoistAcyclic $
+  plam $
+    flip pmatch $ \case
+      PWithin -> pconstant True
+      _ -> pconstant False
 
-{- | True if the 'PProposalTime' is in the locking period.
+{- | Represent a proposal period.
 
-     @since 0.1.0
+     @since 1.0.0
 -}
-isLockingPeriod ::
-  forall (s :: S).
-  Term
-    s
-    ( PProposalTimingConfig
-        :--> PProposalStartingTime
-        :--> PProposalTime
-        :--> PBool
+data PPeriod (s :: S)
+  = PDraftingPeriod
+  | PVotingPeriod
+  | PLockingPeriod
+  | PExecutingPeriod
+  deriving stock
+    ( -- | @since 1.0.0
+      Generic
+    , -- | @since 1.0.0
+      Enum
+    , -- | @since 1.0.0
+      Bounded
+    )
+  deriving anyclass
+    ( -- | @since 1.0.0
+      PlutusType
     )
-isLockingPeriod = phoistAcyclic $
-  plam $ \config s' -> pmatch s' $ \(PProposalStartingTime s) ->
-    pletFields @'["draftTime", "votingTime", "lockingTime"] config $ \f ->
-      pisCurrentTimeWithin # s # (s + f.draftTime + f.votingTime + f.lockingTime)
 
-{- | True if the 'PProposalTime' is in the execution period.
+-- | @since 1.0.0
+instance DerivePlutusType PPeriod where
+  type DPTStrat _ = PlutusTypeEnumData
 
-     @since 0.1.0
+{- | Compute the relation between current time range and the given peroid,
+     providing the starting time and timing configuration of a proposal. If the
+     relation cannot be determined, error out.
+
+     @since 1.0.0
 -}
-isExecutionPeriod ::
+pgetRelation ::
   forall (s :: S).
   Term
     s
     ( PProposalTimingConfig
         :--> PProposalStartingTime
         :--> PProposalTime
-        :--> PBool
+        :--> PPeriod
+        :--> PTimingRelation
     )
-isExecutionPeriod = phoistAcyclic $
-  plam $ \config s' -> pmatch s' $ \(PProposalStartingTime s) ->
-    pletFields @'["draftTime", "votingTime", "lockingTime", "executingTime"] config $ \f ->
-      pisCurrentTimeWithin # s
-        # (s + f.draftTime + f.votingTime + f.lockingTime + f.executingTime)
+pgetRelation = phoistAcyclic $
+  plam $ \config startingTime currentTime period -> unTermCont $ do
+    configF <- pletAllC config
+
+    PProposalStartingTime s <- pmatchC startingTime
+    PCurrentTime lb ub <- pmatchC currentTime
+
+    dub <- pletC $ s + configF.draftTime
+    vub <- pletC $ dub + configF.votingTime
+    lub <- pletC $ vub + configF.lockingTime
+    eub <- pletC $ lub + configF.executingTime
+
+    (plb, pub) <-
+      pmatchC period
+        <&> ( \case
+                PDraftingPeriod -> (s, dub)
+                PVotingPeriod -> (dub, vub)
+                PLockingPeriod -> (vub, lub)
+                PExecutingPeriod -> (lub, eub)
+            )
+
+    pure $
+      pif (plb #<= lb #&& ub #<= pub) (pcon PWithin) $
+        pif (pub #< lb) (pcon PAfter) $
+          ptraceError "pgetRelation: too early or invalid current time"
```

Comment:
This function is great but it is slightly inefficient because the way it is used, we actually always know which `PPeriod` we care about, but we always still compute for all four. This could be fixed if we took it at the Haskell-level. On the other hand, it does also result in less code size currently, so the trade off isn't so clear. Since it is so simple as is, it's probably fine to just leave a comment about this for now.